lib: smf: add API to get the current executing state

ysway opened the pull request:
Expose the current executing state when calling `smf_execute_all_exit_actions`.
This maintains consistency with other context functions and helps when multiple
states share common exit actions. For example, when using wrapper layers on
top of SMF for code compatibility, knowing the currently executing state
becomes important.

ysway committed:
lib: smf: add const qualifier to ctx arg

Add const qualifier to ctx arguments
for immutability and consistency with other functions.

Signed-off-by: Siwei Yang <yswysc@gmail.com>

ysway committed:
lib: smf: add API to get the current leaf and executing state

Retrieve the current leaf state by calling `smf_get_current_leaf_state`
and the state which is currently-executing by calling
`smf_get_current_executing_state`.

Signed-off-by: Siwei Yang <yswysc@gmail.com>

glenn-andrews commented on lib/smf/smf.c:
```diff
@@ -190,10 +190,12 @@ static bool smf_execute_ancestor_run_actions(struct smf_ctx *ctx)
 	for (const struct smf_state *to_execute = ctx->current;
 	     to_execute != NULL && to_execute != topmost; to_execute = to_execute->parent) {
 		if (to_execute->exit) {
+			ctx->executing = to_execute;
```
If the function exits on line 203, `ctx->executing` is not reset.

Now, the state machine has terminated at this point so it should be okay, unless you want to know which state it was in when terminated, in which case it will give an invalid result.

Consider resetting the value on line 203.

ysway replied:
Agreed. It's better to update the relevant information in the context, even if the state machine has been terminated.

Logic added.

glenn-andrews commented:
You've got a compliance check failure to fix.

glenn-andrews reviewed (dismissed):
I'm still not keen on the idea of using internal context to determine external events, but since you are, may as well make it consistent.

keith-zephyr reviewed (commented):
I have the same concerns as @glenn-andrews about SMF user code relying on internal bookkeeping information.

What about adding another public function to the API? This would help insulate SMF users from changes to the SMF library.

ysway commented:
> > > I have the same concerns as @glenn-andrews about SMF user code relying on internal bookkeeping information.
> > > What about adding another public function to the API? This would help insulate SMF users from changes to the SMF library.
> > 
> > 
> > How about adding something like the following to the header? To help the client obtain the state that is being accessed.
> > ```c
> > static inline const struct smf_state* smf_get_executing(const struct smf_ctx *const ctx)
> > {
> > #ifdef CONFIG_SMF_ANCESTOR_SUPPORT
> > 	return ctx->executing;
> > #else
> > 	return ctx->current;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, to make the helper work, the source code still needs to be modified.
> 
> Yes, I think this is a good approach.

Thanks.

Helper added.

glenn-andrews commented:
Now you've changed the interface, add it to the documentation and tests, please, to confirm it works.

ysway commented on include/zephyr/smf.h:
```diff
@@ -142,6 +142,20 @@ struct smf_ctx {
+	return ctx->executing;
+#else
+	return ctx->current;
+}
```
Oh no😂 I'll fix this

ysway commented:
> Now you've changed the interface, add it to the documentation and tests, please, to confirm it works.

Thanks, will do

ysway commented on include/zephyr/smf.h:
```diff
@@ -142,6 +142,20 @@ struct smf_ctx {
+ */
+static inline const struct smf_state* smf_get_executing(const struct smf_ctx *const ctx)
+{
+#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
```
Ok, I'll use `if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT))`

ysway replied:
Tried `if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT))` but didn't work...

The member `executing` of `smf_ctx` is not available when `CONFIG_SMF_ANCESTOR_SUPPORT` is not defiend

ysway commented:
Hi @glenn-andrews, I renamed the title of the MR, added some more code related to `ctx->executing` and edited the commit message and its description.

Since there'll be a new helper to retrieve the current executing state, to meet the description of it, I went through the code which uses `ctx->executing` and added some more logic to make it tracked.

glenn-andrews reviewed (commented):
I'm concerned about the terminology being confusing between 'current' and 'executing' states. I propose 'leaf' and 'executing' state, but am open to more descriptive names.

glenn-andrews commented on doc/services/smf/index.rst:
```diff
@@ -130,6 +130,17 @@ should be called. It can be called from the entry, run, or exit actions. The
+Retrieve the Current  State
+====================================
+
+To retrieve the current state, the :c:func:`smf_get_current` function should be called.
```
Good point! To be honest that leads to malformed HSMs when these are not defined, and we should make it clear in the documentation they should always be set - but that's not the job of this PR.

I think calling it 'leaf state' still works, with the caveats you mentioned written in the documentation here and in the header file.

ysway commented on include/zephyr/smf.h:
```diff
@@ -180,6 +180,32 @@ void smf_set_terminate(struct smf_ctx *ctx, int32_t val);
+ * @param ctx State machine context
+ * @return    The current state.
+ */
+static inline const struct smf_state *smf_get_current(const struct smf_ctx *const ctx)
```
renamed

ysway commented on doc/services/smf/index.rst:
```diff
@@ -130,6 +130,17 @@ should be called. It can be called from the entry, run, or exit actions. The
 function takes a non-zero user defined value that will be returned by the
 :c:func:`smf_run_state` function.
 
+Retrieve the Current  State
```
Fixed

ysway commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -80,6 +80,11 @@ static struct test_object {
 static void state_a_entry(void *obj)
 {
+	zassert_equal(smf_get_executing(SMF_CTX(obj)), &test_states[STATE_A],
+				"Fail to get the current executing state at entry. Expected: State A");
```
Fixed

glenn-andrews reviewed (dismissed):
Looks good. Thank you.

keith-zephyr commented:
Please rebase and fix the compliance errors.  The twister failure looks unrelated to this PR.

ysway commented:
@keith-zephyr Could you re-run the pipeline job https://github.com/zephyrproject-rtos/zephyr/actions/runs/18126394711/job/51582615181?pr=95466 for me? From the error log, I think a re-run may fix it. Thanks.

```
Run python3 /home/runner/work/_actions/zephyrproject-rtos/action-manifest/09983f53d3d878791aa37a7755ae44d695f4c1e5/action.py -p "west.yml" \
/home/runner/work/_actions/zephyrproject-rtos/action-manifest/09983f53d3d878791aa37a7755ae44d695f4c1e5/action.py:457: DeprecationWarning: Argument login_or_token is deprecated, please use auth=github.Auth.Token(...) instead
  gh = Github(token)
ERROR: 'git merge-base 4a69c1b8a8e92d6e77d139ce5dda41cc2012120e 32aee0c595c91072e3033ec2e661e67294ee0b5b' exited with status 128 and/or wrote
to stderr.
==stdout==

==stderr==
fatal: Not a valid commit name 32aee0c595c91072e3033ec2e661e67294ee0b5b
```

glenn-andrews commented:
> ==stderr==
> fatal: Not a valid commit name 32aee0c595c91072e3033ec2e661e67294ee0b5b
> ```

I reran it and got the same error.

ysway commented:
> > ==stderr==
> > fatal: Not a valid commit name [32aee0c](https://github.com/zephyrproject-rtos/zephyr/commit/32aee0c595c91072e3033ec2e661e67294ee0b5b)
> 
> I reran it and got the same error.

That's weird, I'll rebase it with the latest main and push again, hopefully the new has can pass the pipeline🥲

ysway commented:
@glenn-andrews I've rebased and forcefully pushed; the action passed now. ✔️

ysway commented:
@glenn-andrews @keith-zephyr I rebased and pushed again, added the previously missing document in `doc/releases/release-notes-4.3.rst`.
